// carrot-sdk/dist/carrot-sdk.esm.js
import invariant from 'tiny-invariant';
import { BigNumber } from '@ethersproject/bignumber';
import Decimal$1, { Decimal } from 'decimal.js-light';
import { isAddress, getAddress } from '@ethersproject/address';

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;
  subClass.__proto__ = superClass;
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

var Fraction = /*#__PURE__*/function () {
  function Fraction(numerator, denominator) {
    this.numerator = numerator;
    this.denominator = denominator;
  }

  var _proto = Fraction.prototype;

  _proto.add = function add(other) {
    if (this.denominator.eq(other.denominator)) {
      return new Fraction(this.numerator.add(other.numerator), this.denominator);
    }

    return new Fraction(this.numerator.mul(other.denominator).add(other.numerator.mul(this.denominator)), this.denominator.mul(other.denominator));
  };

  _proto.subtract = function subtract(other) {
    if (this.denominator.eq(other.denominator)) {
      return new Fraction(this.numerator.sub(other.numerator), this.denominator);
    }

    return new Fraction(this.numerator.mul(other.denominator).sub(other.numerator.mul(this.denominator)), this.denominator.mul(other.denominator));
  };

  _proto.lessThan = function lessThan(other) {
    return this.numerator.mul(other.denominator).lt(other.numerator.mul(this.denominator));
  };

  _proto.equalTo = function equalTo(other) {
    return this.numerator.mul(other.denominator).eq(other.numerator.mul(this.denominator));
  };

  _proto.greaterThan = function greaterThan(other) {
    return this.numerator.mul(other.denominator).gt(other.numerator.mul(this.denominator));
  };

  _proto.isZero = function isZero() {
    return this.numerator.eq(0);
  };

  _proto.multiply = function multiply(other) {
    return new Fraction(this.numerator.mul(other.numerator), this.denominator.mul(other.denominator));
  };

  _proto.divide = function divide(other) {
    return new Fraction(this.numerator.mul(other.denominator), this.denominator.mul(other.numerator));
  };

  _proto.toFixed = function toFixed(decimalPlaces) {
    !Number.isInteger(decimalPlaces) ? process.env.NODE_ENV !== "production" ? invariant(false, decimalPlaces + " is not an integer.") : invariant(false) : void 0;
    !(decimalPlaces >= 0) ? process.env.NODE_ENV !== "production" ? invariant(false, decimalPlaces + " is negative.") : invariant(false) : void 0;
    return new Decimal(this.numerator.toString()).dividedBy(this.denominator.toString()).toFixed(decimalPlaces);
  };

  return Fraction;
}();

var CurrencyAmount = /*#__PURE__*/function (_Fraction) {
  _inheritsLoose(CurrencyAmount, _Fraction);

  function CurrencyAmount(currency, amount) {
    var _this;

    if (amount instanceof Fraction) {
      _this = _Fraction.call(this, amount.numerator, amount.denominator) || this;
    } else {
      _this = _Fraction.call(this, amount, BigNumber.from(10).pow(currency.decimals)) || this;
    }

    _this.currency = currency;
    return _assertThisInitialized(_this);
  }

  var _proto = CurrencyAmount.prototype;

  _proto.plus = function plus(other) {
    !this.currency.equals(other.currency) ? process.env.NODE_ENV !== "production" ? invariant(false, 'tried to sum different currencies') : invariant(false) : void 0;
    return new CurrencyAmount(this.currency, _Fraction.prototype.add.call(this, other));
  };

  _proto.minus = function minus(other) {
    !this.currency.equals(other.currency) ? process.env.NODE_ENV !== "production" ? invariant(false, 'tried to subtract different currencies') : invariant(false) : void 0;
    return new CurrencyAmount(this.currency, _Fraction.prototype.subtract.call(this, other));
  };

  _proto.multiply = function multiply(other) {
    return new CurrencyAmount(other.currency, _Fraction.prototype.multiply.call(this, other));
  };

  return CurrencyAmount;
}(Fraction);
var TokenAmount = /*#__PURE__*/function (_CurrencyAmount) {
  _inheritsLoose(TokenAmount, _CurrencyAmount);

  function TokenAmount(token, amount) {
    var _this2;

    _this2 = _CurrencyAmount.call(this, token, amount) || this;
    _this2.token = token;
    return _this2;
  }

  return TokenAmount;
}(CurrencyAmount);

var factory = [
	{
		inputs: [
			{
				internalType: "address",
				name: "_kpiTokenImplementation",
				type: "address"
			},
			{
				internalType: "address",
				name: "_oracle",
				type: "address"
			},
			{
				internalType: "address",
				name: "_arbitrator",
				type: "address"
			},
			{
				internalType: "uint16",
				name: "_fee",
				type: "uint16"
			},
			{
				internalType: "uint32",
				name: "_voteTimeout",
				type: "uint32"
			},
			{
				internalType: "address",
				name: "_feeReceiver",
				type: "address"
			}
		],
		stateMutability: "nonpayable",
		type: "constructor"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: false,
				internalType: "address",
				name: "kpiToken",
				type: "address"
			},
			{
				indexed: false,
				internalType: "uint256",
				name: "feeAmount",
				type: "uint256"
			},
			{
				indexed: false,
				internalType: "uint32",
				name: "kpiExpiry",
				type: "uint32"
			}
		],
		name: "KpiTokenCreated",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				internalType: "address",
				name: "previousOwner",
				type: "address"
			},
			{
				indexed: true,
				internalType: "address",
				name: "newOwner",
				type: "address"
			}
		],
		name: "OwnershipTransferred",
		type: "event"
	},
	{
		inputs: [
		],
		name: "arbitrator",
		outputs: [
			{
				internalType: "address",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "string",
				name: "_kpiQuestion",
				type: "string"
			},
			{
				internalType: "uint32",
				name: "_kpiExpiry",
				type: "uint32"
			},
			{
				components: [
					{
						internalType: "address",
						name: "token",
						type: "address"
					},
					{
						internalType: "uint256",
						name: "amount",
						type: "uint256"
					}
				],
				internalType: "struct IKPIToken.Collateral",
				name: "_collateral",
				type: "tuple"
			},
			{
				components: [
					{
						internalType: "string",
						name: "name",
						type: "string"
					},
					{
						internalType: "string",
						name: "symbol",
						type: "string"
					},
					{
						internalType: "uint256",
						name: "totalSupply",
						type: "uint256"
					}
				],
				internalType: "struct IKPIToken.TokenData",
				name: "_tokenData",
				type: "tuple"
			},
			{
				components: [
					{
						internalType: "uint256",
						name: "lowerBound",
						type: "uint256"
					},
					{
						internalType: "uint256",
						name: "higherBound",
						type: "uint256"
					}
				],
				internalType: "struct IKPIToken.ScalarData",
				name: "_scalarData",
				type: "tuple"
			}
		],
		name: "createKpiToken",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
		],
		name: "fee",
		outputs: [
			{
				internalType: "uint16",
				name: "",
				type: "uint16"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "feeReceiver",
		outputs: [
			{
				internalType: "address",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "kpiTokenImplementation",
		outputs: [
			{
				internalType: "address",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "oracle",
		outputs: [
			{
				internalType: "contract IReality",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "owner",
		outputs: [
			{
				internalType: "address",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "renounceOwnership",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "_arbitrator",
				type: "address"
			}
		],
		name: "setArbitrator",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "uint16",
				name: "_fee",
				type: "uint16"
			}
		],
		name: "setFee",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "_feeReceiver",
				type: "address"
			}
		],
		name: "setFeeReceiver",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "uint32",
				name: "_voteTimeout",
				type: "uint32"
			}
		],
		name: "setVoteTimeout",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "newOwner",
				type: "address"
			}
		],
		name: "transferOwnership",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "_kpiTokenImplementation",
				type: "address"
			}
		],
		name: "upgradeKpiTokenImplementation",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
		],
		name: "voteTimeout",
		outputs: [
			{
				internalType: "uint32",
				name: "",
				type: "uint32"
			}
		],
		stateMutability: "view",
		type: "function"
	}
];

var erc20 = [
	{
		constant: true,
		inputs: [
		],
		name: "decimals",
		outputs: [
			{
				name: "",
				type: "uint8"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		constant: true,
		inputs: [
		],
		name: "symbol",
		outputs: [
			{
				name: "",
				type: "string"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		constant: true,
		inputs: [
		],
		name: "name",
		outputs: [
			{
				name: "",
				type: "string"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		constant: true,
		inputs: [
			{
				name: "",
				type: "address"
			}
		],
		name: "balanceOf",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	}
];

var reality = [
	{
		constant: false,
		inputs: [
		],
		name: "withdraw",
		outputs: [
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function",
		signature: "0x3ccfd60b"
	},
	{
		constant: true,
		inputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		name: "template_hashes",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x4e60f883"
	},
	{
		constant: true,
		inputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		name: "question_claims",
		outputs: [
			{
				name: "payee",
				type: "address"
			},
			{
				name: "last_bond",
				type: "uint256"
			},
			{
				name: "queued_funds",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x590158a7"
	},
	{
		constant: true,
		inputs: [
			{
				name: "",
				type: "address"
			}
		],
		name: "arbitrator_question_fees",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x6fa42742"
	},
	{
		constant: true,
		inputs: [
			{
				name: "",
				type: "address"
			}
		],
		name: "balanceOf",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x70a08231"
	},
	{
		constant: true,
		inputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		name: "commitments",
		outputs: [
			{
				name: "reveal_ts",
				type: "uint32"
			},
			{
				name: "is_revealed",
				type: "bool"
			},
			{
				name: "revealed_answer",
				type: "bytes32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x839df945"
	},
	{
		constant: true,
		inputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		name: "questions",
		outputs: [
			{
				name: "content_hash",
				type: "bytes32"
			},
			{
				name: "arbitrator",
				type: "address"
			},
			{
				name: "opening_ts",
				type: "uint32"
			},
			{
				name: "timeout",
				type: "uint32"
			},
			{
				name: "finalize_ts",
				type: "uint32"
			},
			{
				name: "is_pending_arbitration",
				type: "bool"
			},
			{
				name: "bounty",
				type: "uint256"
			},
			{
				name: "best_answer",
				type: "bytes32"
			},
			{
				name: "history_hash",
				type: "bytes32"
			},
			{
				name: "bond",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x95addb90"
	},
	{
		constant: true,
		inputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		name: "templates",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0xbc525652"
	},
	{
		inputs: [
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "constructor",
		signature: "constructor"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: false,
				name: "arbitrator",
				type: "address"
			},
			{
				indexed: false,
				name: "amount",
				type: "uint256"
			}
		],
		name: "LogSetQuestionFee",
		type: "event",
		signature: "0xdca703d022171824d3d639b33c1525fd2338120b4cfb89507c0b59596893acda"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "template_id",
				type: "uint256"
			},
			{
				indexed: true,
				name: "user",
				type: "address"
			},
			{
				indexed: false,
				name: "question_text",
				type: "string"
			}
		],
		name: "LogNewTemplate",
		type: "event",
		signature: "0xb87fb721c0a557bb8dff89a86796466931d82ba530a66a239263eb8735ade2e4"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "question_id",
				type: "bytes32"
			},
			{
				indexed: true,
				name: "user",
				type: "address"
			},
			{
				indexed: false,
				name: "template_id",
				type: "uint256"
			},
			{
				indexed: false,
				name: "question",
				type: "string"
			},
			{
				indexed: true,
				name: "content_hash",
				type: "bytes32"
			},
			{
				indexed: false,
				name: "arbitrator",
				type: "address"
			},
			{
				indexed: false,
				name: "timeout",
				type: "uint32"
			},
			{
				indexed: false,
				name: "opening_ts",
				type: "uint32"
			},
			{
				indexed: false,
				name: "nonce",
				type: "uint256"
			},
			{
				indexed: false,
				name: "created",
				type: "uint256"
			}
		],
		name: "LogNewQuestion",
		type: "event",
		signature: "0xfe2dac156a3890636ce13f65f4fdf41dcaee11526e4a5374531572d92194796c"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "question_id",
				type: "bytes32"
			},
			{
				indexed: false,
				name: "bounty_added",
				type: "uint256"
			},
			{
				indexed: false,
				name: "bounty",
				type: "uint256"
			},
			{
				indexed: true,
				name: "user",
				type: "address"
			}
		],
		name: "LogFundAnswerBounty",
		type: "event",
		signature: "0x54d68405b79f2aa4fd4e8db7b67844ad254cf8f208aac476c2894134a9deab66"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: false,
				name: "answer",
				type: "bytes32"
			},
			{
				indexed: true,
				name: "question_id",
				type: "bytes32"
			},
			{
				indexed: false,
				name: "history_hash",
				type: "bytes32"
			},
			{
				indexed: true,
				name: "user",
				type: "address"
			},
			{
				indexed: false,
				name: "bond",
				type: "uint256"
			},
			{
				indexed: false,
				name: "ts",
				type: "uint256"
			},
			{
				indexed: false,
				name: "is_commitment",
				type: "bool"
			}
		],
		name: "LogNewAnswer",
		type: "event",
		signature: "0xe47ca4ebbbc2990134d1168821f38c5e177f3d5ee564bffeadeaa351905e6221"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "question_id",
				type: "bytes32"
			},
			{
				indexed: true,
				name: "user",
				type: "address"
			},
			{
				indexed: true,
				name: "answer_hash",
				type: "bytes32"
			},
			{
				indexed: false,
				name: "answer",
				type: "bytes32"
			},
			{
				indexed: false,
				name: "nonce",
				type: "uint256"
			},
			{
				indexed: false,
				name: "bond",
				type: "uint256"
			}
		],
		name: "LogAnswerReveal",
		type: "event",
		signature: "0xa7b2d313bc7a062e30b2c3b811aa4c9faf09755a6b4ea3bf42deff920944332f"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "question_id",
				type: "bytes32"
			},
			{
				indexed: true,
				name: "user",
				type: "address"
			}
		],
		name: "LogNotifyOfArbitrationRequest",
		type: "event",
		signature: "0x75d7939999bc902187c4aed400872883e445145f1983539166f783fa040b4762"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "question_id",
				type: "bytes32"
			},
			{
				indexed: true,
				name: "answer",
				type: "bytes32"
			}
		],
		name: "LogFinalize",
		type: "event",
		signature: "0x18d760beffe3717270cd90d9d920ec1a48c194e9ad7bba23eb1c92d3eb974f97"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "question_id",
				type: "bytes32"
			},
			{
				indexed: true,
				name: "user",
				type: "address"
			},
			{
				indexed: false,
				name: "amount",
				type: "uint256"
			}
		],
		name: "LogClaim",
		type: "event",
		signature: "0x9c121aff33b50c1a53fef034ebec5f83da2d5a5187048f9c76c397ba27c1a1a6"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "user",
				type: "address"
			},
			{
				indexed: false,
				name: "amount",
				type: "uint256"
			}
		],
		name: "LogWithdraw",
		type: "event",
		signature: "0x4ce7033d118120e254016dccf195288400b28fc8936425acd5f17ce2df3ab708"
	},
	{
		constant: false,
		inputs: [
			{
				name: "fee",
				type: "uint256"
			}
		],
		name: "setQuestionFee",
		outputs: [
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function",
		signature: "0x4df6ca2a"
	},
	{
		constant: false,
		inputs: [
			{
				name: "content",
				type: "string"
			}
		],
		name: "createTemplate",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function",
		signature: "0x83bf4609"
	},
	{
		constant: false,
		inputs: [
			{
				name: "content",
				type: "string"
			},
			{
				name: "question",
				type: "string"
			},
			{
				name: "arbitrator",
				type: "address"
			},
			{
				name: "timeout",
				type: "uint32"
			},
			{
				name: "opening_ts",
				type: "uint32"
			},
			{
				name: "nonce",
				type: "uint256"
			}
		],
		name: "createTemplateAndAskQuestion",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: true,
		stateMutability: "payable",
		type: "function",
		signature: "0xa1130d04"
	},
	{
		constant: false,
		inputs: [
			{
				name: "template_id",
				type: "uint256"
			},
			{
				name: "question",
				type: "string"
			},
			{
				name: "arbitrator",
				type: "address"
			},
			{
				name: "timeout",
				type: "uint32"
			},
			{
				name: "opening_ts",
				type: "uint32"
			},
			{
				name: "nonce",
				type: "uint256"
			}
		],
		name: "askQuestion",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: true,
		stateMutability: "payable",
		type: "function",
		signature: "0x762c38fd"
	},
	{
		constant: false,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "fundAnswerBounty",
		outputs: [
		],
		payable: true,
		stateMutability: "payable",
		type: "function",
		signature: "0x59245ff3"
	},
	{
		constant: false,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			},
			{
				name: "answer",
				type: "bytes32"
			},
			{
				name: "max_previous",
				type: "uint256"
			}
		],
		name: "submitAnswer",
		outputs: [
		],
		payable: true,
		stateMutability: "payable",
		type: "function",
		signature: "0x77f325df"
	},
	{
		constant: false,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			},
			{
				name: "answer_hash",
				type: "bytes32"
			},
			{
				name: "max_previous",
				type: "uint256"
			},
			{
				name: "_answerer",
				type: "address"
			}
		],
		name: "submitAnswerCommitment",
		outputs: [
		],
		payable: true,
		stateMutability: "payable",
		type: "function",
		signature: "0xd7cff986"
	},
	{
		constant: false,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			},
			{
				name: "answer",
				type: "bytes32"
			},
			{
				name: "nonce",
				type: "uint256"
			},
			{
				name: "bond",
				type: "uint256"
			}
		],
		name: "submitAnswerReveal",
		outputs: [
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function",
		signature: "0x4dc266b4"
	},
	{
		constant: false,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			},
			{
				name: "requester",
				type: "address"
			},
			{
				name: "max_previous",
				type: "uint256"
			}
		],
		name: "notifyOfArbitrationRequest",
		outputs: [
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function",
		signature: "0xf6a94ecb"
	},
	{
		constant: false,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			},
			{
				name: "answer",
				type: "bytes32"
			},
			{
				name: "answerer",
				type: "address"
			}
		],
		name: "submitAnswerByArbitrator",
		outputs: [
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function",
		signature: "0xfe92049d"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "isFinalized",
		outputs: [
			{
				name: "",
				type: "bool"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x7f8d429e"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getFinalAnswer",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0xa462fb7b"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "resultFor",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0xd09cc57e"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			},
			{
				name: "content_hash",
				type: "bytes32"
			},
			{
				name: "arbitrator",
				type: "address"
			},
			{
				name: "min_timeout",
				type: "uint32"
			},
			{
				name: "min_bond",
				type: "uint256"
			}
		],
		name: "getFinalAnswerIfMatches",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x12a203c3"
	},
	{
		constant: false,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			},
			{
				name: "history_hashes",
				type: "bytes32[]"
			},
			{
				name: "addrs",
				type: "address[]"
			},
			{
				name: "bonds",
				type: "uint256[]"
			},
			{
				name: "answers",
				type: "bytes32[]"
			}
		],
		name: "claimWinnings",
		outputs: [
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function",
		signature: "0x1101a0fd"
	},
	{
		constant: false,
		inputs: [
			{
				name: "question_ids",
				type: "bytes32[]"
			},
			{
				name: "lengths",
				type: "uint256[]"
			},
			{
				name: "hist_hashes",
				type: "bytes32[]"
			},
			{
				name: "addrs",
				type: "address[]"
			},
			{
				name: "bonds",
				type: "uint256[]"
			},
			{
				name: "answers",
				type: "bytes32[]"
			}
		],
		name: "claimMultipleAndWithdrawBalance",
		outputs: [
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function",
		signature: "0x28828b1e"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getContentHash",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x51577ea9"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getArbitrator",
		outputs: [
			{
				name: "",
				type: "address"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x2518904c"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getOpeningTS",
		outputs: [
			{
				name: "",
				type: "uint32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x9e63fa6a"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getTimeout",
		outputs: [
			{
				name: "",
				type: "uint32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x9f1025c6"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getFinalizeTS",
		outputs: [
			{
				name: "",
				type: "uint32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0xacae8f4e"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "isPendingArbitration",
		outputs: [
			{
				name: "",
				type: "bool"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x924532fb"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getBounty",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x2417395c"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getBestAnswer",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x8d552d46"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getHistoryHash",
		outputs: [
			{
				name: "",
				type: "bytes32"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x82ffa9f7"
	},
	{
		constant: true,
		inputs: [
			{
				name: "question_id",
				type: "bytes32"
			}
		],
		name: "getBond",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function",
		signature: "0x26d6c97b"
	}
];

var kpiToken = [
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				internalType: "address",
				name: "owner",
				type: "address"
			},
			{
				indexed: true,
				internalType: "address",
				name: "spender",
				type: "address"
			},
			{
				indexed: false,
				internalType: "uint256",
				name: "value",
				type: "uint256"
			}
		],
		name: "Approval",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: false,
				internalType: "uint256",
				name: "finalKpiProgress",
				type: "uint256"
			}
		],
		name: "Finalized",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: false,
				internalType: "bytes32",
				name: "kpiId",
				type: "bytes32"
			},
			{
				indexed: false,
				internalType: "address",
				name: "oracle",
				type: "address"
			},
			{
				indexed: false,
				internalType: "address",
				name: "creator",
				type: "address"
			},
			{
				components: [
					{
						internalType: "address",
						name: "token",
						type: "address"
					},
					{
						internalType: "uint256",
						name: "amount",
						type: "uint256"
					}
				],
				indexed: false,
				internalType: "struct IKPIToken.Collateral",
				name: "collateral",
				type: "tuple"
			},
			{
				components: [
					{
						internalType: "string",
						name: "name",
						type: "string"
					},
					{
						internalType: "string",
						name: "symbol",
						type: "string"
					},
					{
						internalType: "uint256",
						name: "totalSupply",
						type: "uint256"
					}
				],
				indexed: false,
				internalType: "struct IKPIToken.TokenData",
				name: "tokenData",
				type: "tuple"
			},
			{
				components: [
					{
						internalType: "uint256",
						name: "lowerBound",
						type: "uint256"
					},
					{
						internalType: "uint256",
						name: "higherBound",
						type: "uint256"
					}
				],
				indexed: false,
				internalType: "struct IKPIToken.ScalarData",
				name: "scalarData",
				type: "tuple"
			}
		],
		name: "Initialized",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: false,
				internalType: "uint256",
				name: "burnedTokens",
				type: "uint256"
			},
			{
				indexed: false,
				internalType: "uint256",
				name: "redeemedCollateral",
				type: "uint256"
			}
		],
		name: "Redeemed",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				internalType: "address",
				name: "from",
				type: "address"
			},
			{
				indexed: true,
				internalType: "address",
				name: "to",
				type: "address"
			},
			{
				indexed: false,
				internalType: "uint256",
				name: "value",
				type: "uint256"
			}
		],
		name: "Transfer",
		type: "event"
	},
	{
		inputs: [
		],
		name: "INVALID_ANSWER",
		outputs: [
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "owner",
				type: "address"
			},
			{
				internalType: "address",
				name: "spender",
				type: "address"
			}
		],
		name: "allowance",
		outputs: [
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "spender",
				type: "address"
			},
			{
				internalType: "uint256",
				name: "amount",
				type: "uint256"
			}
		],
		name: "approve",
		outputs: [
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "account",
				type: "address"
			}
		],
		name: "balanceOf",
		outputs: [
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "collateralToken",
		outputs: [
			{
				internalType: "contract IERC20Upgradeable",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "creator",
		outputs: [
			{
				internalType: "address",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "decimals",
		outputs: [
			{
				internalType: "uint8",
				name: "",
				type: "uint8"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "spender",
				type: "address"
			},
			{
				internalType: "uint256",
				name: "subtractedValue",
				type: "uint256"
			}
		],
		name: "decreaseAllowance",
		outputs: [
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
		],
		name: "finalKpiProgress",
		outputs: [
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "finalize",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
		],
		name: "finalized",
		outputs: [
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "spender",
				type: "address"
			},
			{
				internalType: "uint256",
				name: "addedValue",
				type: "uint256"
			}
		],
		name: "increaseAllowance",
		outputs: [
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "bytes32",
				name: "_kpiId",
				type: "bytes32"
			},
			{
				internalType: "address",
				name: "_oracle",
				type: "address"
			},
			{
				internalType: "address",
				name: "_creator",
				type: "address"
			},
			{
				components: [
					{
						internalType: "address",
						name: "token",
						type: "address"
					},
					{
						internalType: "uint256",
						name: "amount",
						type: "uint256"
					}
				],
				internalType: "struct IKPIToken.Collateral",
				name: "_collateral",
				type: "tuple"
			},
			{
				components: [
					{
						internalType: "string",
						name: "name",
						type: "string"
					},
					{
						internalType: "string",
						name: "symbol",
						type: "string"
					},
					{
						internalType: "uint256",
						name: "totalSupply",
						type: "uint256"
					}
				],
				internalType: "struct IKPIToken.TokenData",
				name: "_tokenData",
				type: "tuple"
			},
			{
				components: [
					{
						internalType: "uint256",
						name: "lowerBound",
						type: "uint256"
					},
					{
						internalType: "uint256",
						name: "higherBound",
						type: "uint256"
					}
				],
				internalType: "struct IKPIToken.ScalarData",
				name: "_scalarData",
				type: "tuple"
			}
		],
		name: "initialize",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
		],
		name: "kpiId",
		outputs: [
			{
				internalType: "bytes32",
				name: "",
				type: "bytes32"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "name",
		outputs: [
			{
				internalType: "string",
				name: "",
				type: "string"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "oracle",
		outputs: [
			{
				internalType: "contract IReality",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "redeem",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
		],
		name: "scalarData",
		outputs: [
			{
				internalType: "uint256",
				name: "lowerBound",
				type: "uint256"
			},
			{
				internalType: "uint256",
				name: "higherBound",
				type: "uint256"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "symbol",
		outputs: [
			{
				internalType: "string",
				name: "",
				type: "string"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "totalSupply",
		outputs: [
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "recipient",
				type: "address"
			},
			{
				internalType: "uint256",
				name: "amount",
				type: "uint256"
			}
		],
		name: "transfer",
		outputs: [
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "sender",
				type: "address"
			},
			{
				internalType: "address",
				name: "recipient",
				type: "address"
			},
			{
				internalType: "uint256",
				name: "amount",
				type: "uint256"
			}
		],
		name: "transferFrom",
		outputs: [
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "nonpayable",
		type: "function"
	}
];

var _FACTORY_ADDRESS, _REALITY_ADDRESS;
var ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
var ChainId;

(function (ChainId) {
  ChainId[ChainId["MAINNET"] = 1] = "MAINNET";
  ChainId[ChainId["RINKEBY"] = 4] = "RINKEBY";
})(ChainId || (ChainId = {}));

var FACTORY_ADDRESS = (_FACTORY_ADDRESS = {}, _FACTORY_ADDRESS[ChainId.MAINNET] = ZERO_ADDRESS, _FACTORY_ADDRESS[ChainId.RINKEBY] = '0xa60c831dc30A6564Aa23044Ade594C2f1e3c9929', _FACTORY_ADDRESS);
var REALITY_ADDRESS = (_REALITY_ADDRESS = {}, _REALITY_ADDRESS[ChainId.MAINNET] = '0x325a2e0F3CCA2ddbaeBB4DfC38Df8D19ca165b47', _REALITY_ADDRESS[ChainId.RINKEBY] = '0x3D00D77ee771405628a4bA4913175EcC095538da', _REALITY_ADDRESS);

var Currency = /*#__PURE__*/function () {
  function Currency(decimals, symbol, name) {
    this.decimals = decimals;
    this.symbol = symbol;
    this.name = name;
  }

  Currency.getNative = function getNative(chainId) {
    switch (chainId) {
      case ChainId.RINKEBY:
      case ChainId.MAINNET:
        {
          return Currency.ETHER;
        }

      default:
        {
           process.env.NODE_ENV !== "production" ? invariant(false, 'not a valid chain id when getting native currency') : invariant(false) ;
        }
    }
  };

  var _proto = Currency.prototype;

  _proto.equals = function equals(other) {
    if (this instanceof Token && other instanceof Token) {
      return this.equals(other);
    }

    return this === other;
  };

  return Currency;
}();
Currency.ETHER = /*#__PURE__*/new Currency( /*#__PURE__*/BigNumber.from(18), 'ETH', 'Ether');
var Token = /*#__PURE__*/function (_Currency) {
  _inheritsLoose(Token, _Currency);

  function Token(chainId, address, decimals, symbol, name) {
    var _this;

    _this = _Currency.call(this, decimals, symbol, name) || this;
    _this.chainId = chainId;
    !isAddress(address) ? process.env.NODE_ENV !== "production" ? invariant(false, address + " is not a valid address") : invariant(false) : void 0;
    _this.address = getAddress(address);
    return _this;
  }

  var _proto2 = Token.prototype;

  _proto2.equals = function equals(other) {
    if (this === other) {
      return true;
    }

    return this.chainId === other.chainId && this.address === other.address;
  };

  return Token;
}(Currency);

var KpiToken = /*#__PURE__*/function (_Token) {
  _inheritsLoose(KpiToken, _Token);

  function KpiToken(chainId, address, symbol, name, kpiId, totalSupply, oracle, question, lowerBound, higherBound, finalProgress, expiresAt, finalized, kpiReached, creator, collateral, fee) {
    var _this;

    _this = _Token.call(this, chainId, address, BigNumber.from('18'), symbol, name) || this;
    !(collateral.token.chainId === chainId) ? process.env.NODE_ENV !== "production" ? invariant(false, 'inconsistent chain id in collateral') : invariant(false) : void 0;
    !fee.token.equals(collateral.token) ? process.env.NODE_ENV !== "production" ? invariant(false, 'inconsistent collateral and fee token') : invariant(false) : void 0;
    !(fee.token.chainId === chainId) ? process.env.NODE_ENV !== "production" ? invariant(false, 'inconsistent chain id in collateral') : invariant(false) : void 0;
    !lowerBound.lt(higherBound) ? process.env.NODE_ENV !== "production" ? invariant(false, 'inconsistent scalar bounds') : invariant(false) : void 0;
    _this.kpiId = kpiId;
    _this.totalSupply = new TokenAmount(_assertThisInitialized(_this), totalSupply);
    _this.oracle = oracle;
    _this.question = question;
    _this.lowerBound = lowerBound;
    _this.higherBound = higherBound;
    _this.finalProgress = finalProgress;
    _this.expiresAt = expiresAt;
    _this.finalized = finalized;
    _this.kpiReached = kpiReached;
    _this.creator = creator;
    _this.collateral = collateral;
    _this.fee = fee;
    return _this;
  }

  _createClass(KpiToken, [{
    key: "progressPercentage",
    get: function get() {
      var kpiScalarRange = this.higherBound.sub(this.lowerBound);
      return new Decimal$1(this.finalProgress.toString()).dividedBy(kpiScalarRange.toString()).times(100);
    }
  }]);

  return KpiToken;
}(Token);

export { ChainId, Currency, CurrencyAmount, erc20 as ERC20_ABI, factory as FACTORY_ABI, FACTORY_ADDRESS, kpiToken as KPI_TOKEN_ABI, KpiToken, reality as REALITY_ABI, REALITY_ADDRESS, Token, TokenAmount, ZERO_ADDRESS };
